src/compiler/checker.ts

```diff
@@ -8095,7 +8106,7 @@ namespace ts {
                 }
                 const propName = propAccess.name.text;
                 const propType = getTypeOfPropertyOfType(type, propName);
-                if (!propType || !isStringLiteralUnionType(propType)) {
+                if (!propType || !(isStringLiteralUnionType(propType) || (type.flags & TypeFlags.UnionOrIntersection))) {

```

Comment:
Hm, maybe the better way to write it is:
```ts 
if (!propType || (!isStringLiteralUnionType(propType) && (type.flags & TypeFlags.UnionOrIntersection) === 0))
```
i.e. if neither the discrimination property is a string-literal-union, nor the parent type is union or intersection. I guess I made it unintelligible by applying De Morgan's law.